src/Xamarin.Android.Build.Tasks/Utilities/ApplicationConfigNativeAssemblyGenerator.cs

```diff
@@ -16,8 +21,77 @@ enum MonoComponent
 		Tracing   = 0x04,
 	}
 
-	class ApplicationConfigNativeAssemblyGenerator : NativeAssemblyGenerator
+	class ApplicationConfigNativeAssemblyGenerator : NativeAssemblyComposer
 	{
+		sealed class DSOCacheEntryContextDataProvider : NativeAssemblerStructContextDataProvider
+		{
+			public override string GetComment (object data, string fieldName)
+			{
+				var dso_entry = data as DSOCacheEntry;
+				if (dso_entry == null) {
+					throw new InvalidOperationException ("Invalid data type, expected an instance of DSOCacheEntry");
+				}
+
+				if (String.Compare ("hash", fieldName, StringComparison.Ordinal) == 0) {
+					return $"hash, from name: {dso_entry.HashedName}";
+				}
+
+				if (String.Compare ("name", fieldName, StringComparison.Ordinal) == 0) {
+					return $"name: {dso_entry.name}";
+				}
+
+				return String.Empty;
+			}
+		}
+
+		[NativeAssemblerStructContextDataProvider (typeof (DSOCacheEntryContextDataProvider))]
+		sealed class DSOCacheEntry
+		{
+			[NativeAssembler (Ignore = true)]
+			public string HashedName;
+
+			[NativeAssembler (UsesDataProvider = true)]
+			public ulong hash;
+			public bool ignore;
+
+			[NativeAssemblerString (UsesDataProvider = true)]
+			public string name;
+			public IntPtr handle = IntPtr.Zero;
+		}
+
+		// Order of fields and their type must correspond *exactly* to that in
+		// src/monodroid/jni/xamarin-app.hh AssemblyStoreSingleAssemblyRuntimeData structure
+		sealed class AssemblyStoreSingleAssemblyRuntimeData
+		{
+			public IntPtr  image_data;
+			public IntPtr  debug_info_data;
+			public IntPtr  config_data;
+			public IntPtr  descriptor;
+		};
+
+		// Order of fields and their type must correspond *exactly* to that in
+		// src/monodroid/jni/xamarin-app.hh AssemblyStoreRuntimeData structure
+		sealed class AssemblyStoreRuntimeData
+		{
+			public IntPtr data_start;
+			public uint   assembly_count;
+			public IntPtr assemblies;
+		};
+
+		// Order of fields and their type must correspond *exactly* to that in
+		// src/monodroid/jni/xamarin-app.hh XamarinAndroidBundledAssembly structure
+		sealed class XamarinAndroidBundledAssembly
+		{
+			public int    apk_fd;
+			public uint   data_offset;
+			public uint   data_size;
+			public IntPtr data;
+			public uint   name_length;
+
+			[NativeAssemblerString (PointerToSymbol = true)]
```

Comment:
/me sees "pointer to symbol", and is not sure if he should flee, or *flee very quickly*.

What does this *mean*?

Reply:
I can rename it to `PointerToNativeSymbolInELFBinary` if you think it's better :) I'm not sure there's a need to explain what a symbol in native code is, and what is a pointer to it. I'd love to assume a basic familiarity of underlying technologies of anyone modifying this code. What name would you like to see here?

Reply:
I took a slightly different approach, but similar:

```C#
enum AssemblerStringFormat
{
	/// <summary>
	/// Local string label is automatically generated for the string, this is the default behavior
	/// </summary>
	Automatic,

	/// <summary>
	/// String is output as an inline (no pointer) array/buffer filled directly with string data
	/// </summary>
	InlineArray,

	/// <summary>
	/// String is output as a pointer to symbol whose name is contents of the string being processed
	PointerToSymbol,
}

[AttributeUsage (AttributeTargets.Field | AttributeTargets.Property, Inherited = true)]
class NativeAssemblerStringAttribute : NativeAssemblerAttribute
{
	AssemblerStringFormat format;

	public bool Inline          => format == AssemblerStringFormat.InlineArray;
	public bool PadToMaxLength  { get; set; }
	public bool PointerToSymbol => format == AssemblerStringFormat.PointerToSymbol;

	public NativeAssemblerStringAttribute ()
	{
		format = AssemblerStringFormat.Automatic;
	}

	public NativeAssemblerStringAttribute (AssemblerStringFormat format)
	{
		this.format = format;
	}
}
```